mixer/pkg/runtime2/routing/builder_test.go

```diff
@@ -0,0 +1,534 @@
+		Name:          "builder-mapper-error-causes-dependent-rules-to-be-omitted",
+		ServiceConfig: data.ServiceConfig,
+		Configs: []string{
+			data.HandlerACheck1,
+			data.InstanceCheck1,
+			data.RuleCheck1,
+		},
+		Templates: data.BuildTemplates(data.FakeTemplateSettings{Name: "tcheck", ErrorAtCreateInstanceBuilder: true}),
+		ExpectedTable: `
+[Routing ExpectedTable]
+ID: 1
+`},
+
+	{
+		Name:          "match-condition-wrong-return-type-in-match-claus-causes-rule-to-be-omitted.",
+		ServiceConfig: data.ServiceConfig,
+		Configs: []string{
+			data.HandlerACheck1,
+			data.InstanceCheck1,
+			data.RuleCheck1WithNonBooleanCondition,
+		},
+
+		ExpectedTable: `
+[Routing ExpectedTable]
+ID: 1
+`},
+
+	{
+		Name:          "apa",
+		ServiceConfig: data.ServiceConfig,
+		Configs: []string{
+			data.HandlerAPA1,
+			data.InstanceAPA1,
+			data.RuleApa1,
+		},
+
+		ExpectedTable: `
+[Routing ExpectedTable]
+ID: 1
+[#0] TEMPLATE_VARIETY_ATTRIBUTE_GENERATOR {V}
+  [#0] istio-system {NS}
+    [#0] hapa1.apa.istio-system {H}
+      [#0]
+        Condition: <NONE>
+        [#0] iapa1.tapa.istio-system {I}
+`},
+
+	{
+		Name:          "apa-error-at-output-expressions",
+		ServiceConfig: data.ServiceConfig,
+		Configs: []string{
+			data.HandlerAPA1,
+			data.InstanceAPA1,
+			data.RuleApa1,
+		},
+		Templates: data.BuildTemplates(data.FakeTemplateSettings{Name: "tapa", ErrorAtCreateOutputExpressions: true}),
+
+		ExpectedTable: `
+[Routing ExpectedTable]
+ID: 1
+`,
+	},
+
+	{
+		Name:          "different-templates-in-same-rule",
+		ServiceConfig: data.ServiceConfig,
+		Configs: []string{
+			data.HandlerACheck1,
+			data.InstanceCheck1,
+			data.InstanceHalt1,
+			data.Rule4CheckAndHalt,
+		},
+
+		ExpectedTable: `
+[Routing ExpectedTable]
+ID: 1
+[#0] TEMPLATE_VARIETY_CHECK {V}
+  [#0] istio-system {NS}
+    [#0] hcheck1.acheck.istio-system {H}
+      [#0]
+        Condition: <NONE>
+        [#0] icheck1.tcheck.istio-system {I}
+    [#1] hcheck1.acheck.istio-system {H}
+      [#0]
+        Condition: <NONE>
+        [#0] ihalt1.thalt.istio-system {I}`,
+	},
+}
+
+func TestBuilder(t *testing.T) {
+	for _, tst := range tests {
+		t.Run(tst.Name, func(tt *testing.T) {
+
+			serviceConfig := tst.ServiceConfig
+			if len(serviceConfig) == 0 {
+				serviceConfig = data.ServiceConfig
+			}
+
+			templates := tst.Templates
+			if templates == nil {
+				templates = data.BuildTemplates()
+			}
+			adapters := tst.Adapters
+			if adapters == nil {
+				adapters = data.BuildAdapters()
+			}
+			t, s := buildTableWithTemplatesAndAdapters(templates, adapters, serviceConfig, tst.Configs, true)
+
+			actual := t.String()
+
+			if normalize(actual) != normalize(tst.ExpectedTable) {
+				tt.Logf("Config:\n%v\n\n", tst.Configs)
+				tt.Logf("Snapshot:\n%s\n\n", s)
+				tt.Logf("Debug: true")
+				tt.Fatalf("%v\n!=\n%v", actual, tst.ExpectedTable)
+			}
+
+			// rerun with debug = false to ensure there is no crash.
+			t, _ = buildTable(serviceConfig, tst.Configs, false)
```

Comment:
should we add a panic handler here to more appropriately log error messages and failures?